### ISSUE TITLE
How about integrating a ECM pre-processing step?

### ISSUE THREAD

**Issue by i30817** [unknownbrackets/maxcso#5]
I dunno specifically how PSP (and ps2 i guess) cds actually work at the physical layer, but ecm works quite fine on ps1 cd images to remove the Error Correction Codes (EEC) from the binary that can be regenerated from the original bytes, therefore being a specialized lossless compression step that helps quite a lot.

Merits investigation? Some emulators load ecm directly (pcsxr) but i guess a additional layer of compression would still be nice, since this is ofc, just removing redundant data, not compressing.

edit: and a post processing step that reverses the thing on output obviously. If you ever implement decompression, which is what i'm obviously missing...

**Comment by unknownbrackets** [comment]
Well, effectively, CSO only supports "ECM" data.  At least, afaiu this is talking about the error correction data, which make sectors generally NOT a power of two in size.  Currently, maxcso and all cso implementations I know of require sectors to be a size that is a power of two.

I think this is highly related to #3, because while DVDs do contain error correction data, I'm not aware that you can actually read this data.  I think only certain DVD readers support this, and I also think this is an issue for Wii discs usually, because they swap they store the error correction data differently.  That being said, they still have standard sectors (just in a different layout.)

In other words, I'm not sure "non-ECM" is a thing outside PS1.  There's also "mode 2" CDs, which do not actually contain ECC data.  You can't remove the ECC data safely from these.  This is as discussed in #3.

-[Unknown]

**Comment by i30817** [comment]
I'd love a new format that was best of breed and incorporated optional steps... (with emulator support naturally, which is the hard part).

Say, a multidisc compressed format, with some standardized optional data headers and room in the specification for adding new data headers as ideas come up.

So, using compression, multidisc support with emulator support for changing them (this is missing on most emulators that claim to support pbp, they always choose the first img) and these headers:
banner image header (used for game list in emulators)
front image / cd case header (maybe used in OS explorers, if they support it - not your problem)
back image / cd case back header
game manual header (in some standard format like pdf if it must)
header for reversible xdelta patch with support for upgrades (for each cd image)
optional ecm pre-compression (but after xdelta patch) step for psx emulation

Associate extensions with consoles for each emulator like "Tekken.psx.cso3"

It would be pretty nice to do away with all of the format crap in one go.

**Comment by unknownbrackets** [comment]
I think this is outside of the scope of maxcso, but yes it does likely make sense.  To some degree, this is what Sony tried to do with PBP, but obviously only for them / the PSP.

I guess what you'd be looking at is a record based format like this:

FileHeader
 - magic
 - version
 - flags / block_size? / etc.
 - part_count

PartDescriptor*
 - type
 - flags
 - size
 - data_offset

[DATA]

Where CSO would just be a subformat of this (e.g. it would be the data for parts of type = ISO1, ISO2, etc.)  This is really similar to e.g. id3/riff/etc., although id3 puts the data right after the header, rather than using an offset.  An offset isn't strictly needed since one can seek, of course.

Anyway, this isn't strictly part of maxcso - what you're asking about is essentially a container format (like e.g. mp4), where the CSO would be the data (like h.264), along with JPG, xdelta, whatever for the various parts.

-[Unknown]
